test/dns-test.js

```diff
@@ -0,0 +1,396 @@
+        // yet use RFC 6979, so nullify the
+        // signatures before comparing them
+
+        if (bcrypto.native === 2)
+          assert.deepEqual(nullSig(res), nullSig(dns));
+        else
+          assert.deepEqual(res.name, dns.name);
+      });
+    }
+  });
+});
+
+/**
+ * Mine blocks and take into
+ * account race conditions
+ */
+
+async function mineBlocks(count, address) {
+  for (let i = 0; i < count; i++) {
+    const obj = { complete: false };
+    node.once('block', () => {
+      obj.complete = true;
+    });
+    await nclient.execute('generatetoaddress', [1, address]);
+    await common.forValue(obj, 'complete', true);
+  }
+}
+
+/**
+ * Verify DNSSEC for each name in the
+ * response. Only check the answer and
+ * authority sections. If all rrsets
+ * are signed, the responses will be
+ * too large and not fit into udp packets
+ */
+
+function verifyDNSSEC(resource, pubkey, qtype, name) {
+  // rr types that are not committed to in DNSSEC
+  const skip = new Set([
+    types.RRSIG
+  ]);
+
+  name = util.fqdn(name);
+
+  const {answer, authority} = resource;
+  const targets = [answer, authority];
+
+  const records = [];
+
+  for (const target of targets) {
+    const toVerify = new Set();
+    for (const record of target) {
+      records.push(record);
+
+      if (!skip.has(record.type))
+        toVerify.add(record.type);
+    }
+
+    for (const type of toVerify) {
+      const rrsig = target.find((rr) => {
+        return rr.type === types.RRSIG
+          && rr.data.typeCovered === type;
+      });
+
+      if (name !== rrsig.name) {
+        const record = records.find((rr) => {
+          return rr.name === util.fqdn(name)
+            && (rr.type === types.CNAME
+            || rr.type === types.DNAME
+            || rr.type === types.SRV
+            || rr.type === types.NS);
+        });
+
+        assert(record);
+
+        switch (qtype) {
+          case types.CNAME:
+          case types.DNAME:
+          case types.SRV:
+            name = record.data.target;
+            break;
+          case types.NS:
+            name = record.data.ns;
+            break;
+          default:
+            assert(false, 'rrsig name does not match');
+        }
+      }
+
+      const rrs = util.extractSet(target, name, type);
+
+      const valid = dnssec.verify(rrsig, pubkey, rrs);
+      assert(valid);
+    }
+  }
+}
+
+/**
+ * Nullify out any signatures in
+ * a DNS response.
+ */
+
+function nullSig(resource) {
+  for (const answer of resource.answer) {
+    if (answer.data.signature)
+      answer.data.signature = null;
+  }
+
+  for (const authority of resource.authority) {
+    if (authority.data.signature)
+      authority.data.signature = null;
+  }
+
+  for (const additional of resource.additional) {
+    if (additional.data.signature)
+      additional.data.signature = null;
+  }
+}
+
+/**
+ * Parse the ZSK and KSK from
+ * a record.
+ */
+
+function getSigningKeys(record) {
```

Comment:
It is possible to do this without calling `rr.toJSON` and looking at the `keyType`, but it is a lot more simple to do it this way when performance isn't critical.

Reply:
Might be a silly question but in cases where there's multiple DNSKEY records (and multiple ZSK/KSK for some reason), this would select the last. Is that fine or maybe we could add a `break`?